workflow/sync/lock_name.go

```diff
@@ -8,91 +8,99 @@ import (
 	"github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
 )
 
-type LockKind string
+type lockKind string
 
 const (
-	LockKindConfigMap LockKind = "ConfigMap"
-	LockKindMutex     LockKind = "Mutex"
+	lockKindConfigMap lockKind = "ConfigMap"
+	lockKindMutex     lockKind = "Mutex"
 )
 
-type LockName struct {
+type lockName struct {
 	Namespace    string
 	ResourceName string
 	Key          string
-	Kind         LockKind
+	Kind         lockKind
 }
 
-func NewLockName(namespace, resourceName, lockKey string, kind LockKind) *LockName {
-	return &LockName{
+func newLockName(namespace, resourceName, lockKey string, kind lockKind) *lockName {
+	return &lockName{
 		Namespace:    namespace,
 		ResourceName: resourceName,
 		Key:          lockKey,
 		Kind:         kind,
 	}
 }
 
-func GetLockName(sync *v1alpha1.Synchronization, wfNamespace string) (*LockName, error) {
-	switch sync.GetType() {
-	case v1alpha1.SynchronizationTypeSemaphore:
-		if sync.Semaphore.ConfigMapKeyRef != nil {
-			namespace := sync.Semaphore.Namespace
-			if namespace == "" {
-				namespace = wfNamespace
-			}
-			return NewLockName(namespace, sync.Semaphore.ConfigMapKeyRef.Name, sync.Semaphore.ConfigMapKeyRef.Key, LockKindConfigMap), nil
-		}
-		return nil, fmt.Errorf("cannot get LockName for a Semaphore without a ConfigMapRef")
-	case v1alpha1.SynchronizationTypeMutex:
-		namespace := sync.Mutex.Namespace
+func getSemaphoreLockName(sem *v1alpha1.SemaphoreRef, wfNamespace string) (*lockName, error) {
+	if sem.ConfigMapKeyRef != nil {
+		namespace := sem.Namespace
 		if namespace == "" {
 			namespace = wfNamespace
 		}
-		return NewLockName(namespace, sync.Mutex.Name, "", LockKindMutex), nil
+		return newLockName(namespace, sem.ConfigMapKeyRef.Name, sem.ConfigMapKeyRef.Key, lockKindConfigMap), nil
+	}
+	return nil, fmt.Errorf("cannot get LockName for a Semaphore without a ConfigMapRef")
+}
+
+func getMutexLockName(mtx *v1alpha1.Mutex, wfNamespace string) *lockName {
+	namespace := mtx.Namespace
+	if namespace == "" {
+		namespace = wfNamespace
+	}
+	return newLockName(namespace, mtx.Name, "", lockKindMutex)
+}
+
+func getLockName(item *syncItem, wfNamespace string) (*lockName, error) {
+	switch {
```

Comment:
SyncItem should have a semaphore or mutex. It needs to be validated in validate.go

Reply:
syncItem is always accessed from syncManager, and always under a scenario where syncManager holds `Manager.lock`. Rather than add more overhead here, I'm relying on that for mutual exclusion.

syncItem is an internal type for sync_manger to manage all synchronization types (mutex and semaphore) as a single type. It's never user exposed and so I can't see how it needs validating.

Reply:
Imo SyncItem does not need anything else here, agree with @Joibel on this one.

Reply:
similar concept was discussed below in https://github.com/argoproj/argo-workflows/pull/13358#discussion_r1714498505. `Unknown` already existed before, but would be good to prevent it entirely during validation as a separate PR